compiler/rustc_middle/src/ty/util.rs

```diff
@@ -1154,6 +1154,14 @@ pub fn normalize_opaque_types<'tcx>(
     val.fold_with(&mut visitor)
 }
 
+/// Determines whether an item is annotated with `doc(hidden)`.
+pub fn is_doc_hidden(tcx: TyCtxt<'_>, def_id: DefId) -> bool {
+    tcx.get_attrs(def_id)
+        .iter()
+        .filter_map(|attr| if attr.has_name(sym::doc) { attr.meta_item_list() } else { None })
+        .any(|items| items.iter().any(|item| item.has_name(sym::hidden)))
+}
+
 pub fn provide(providers: &mut ty::query::Providers) {
-    *providers = ty::query::Providers { normalize_opaque_types, ..*providers }
+    *providers = ty::query::Providers { normalize_opaque_types, is_doc_hidden, ..*providers }
```

Comment:
this only provides it for the local crate I think, you should double check it doesn't crash on external crates

Reply:
Huh, I didn't realize that, thanks!

Reply:
The distinction between local and extern query providers only exists if the query is markes `separate_provide_extern`. This query is not marked as such, so this implementation will be called for both the local crate and extern crates.